### PR TITLE
Add Logging Mechanism

### PR DESCRIPTION
**Pull request by f2prateek**
This is mostly a copy paste from the v1. I didn't want to pass write code to update `LogLevel` everywhere or make it final, so it's folded into `Logger` for now.

### CODE REVIEW

File: retrofit/src/main/java/retrofit/OkHttpCall.java [hunk 1]
```diff
@@ -85,9 +95,10 @@ private void callSuccess(Response<T> response) {
       }
 
       @Override public void onResponse(com.squareup.okhttp.Response rawResponse) {
+        long elapsedTime = TimeUnit.NANOSECONDS.toMillis(System.nanoTime() - start);
```
Comment (artem-zinnatullin) [inline retrofit/src/main/java/retrofit/OkHttpCall.java]: `final`?

File: retrofit/src/main/java/retrofit/Retrofit.java [hunk 2]
```diff
@@ -261,6 +265,12 @@ public Builder validateEagerly() {
       return this;
     }
 
+    /** Configure debug logging mechanism. */
+    public Builder setLogger(Logger logger) {
```
Comment (artem-zinnatullin) [inline retrofit/src/main/java/retrofit/Retrofit.java]: Other methods of the `Builder` don't have `set` prefix, I'd remove it for consistency :)

File: retrofit/src/main/java/retrofit/OkHttpCall.java [hunk 3]
```diff
@@ -109,14 +120,93 @@ private void callSuccess(Response<T> response) {
...
+                    return contentType;
+                  }
+
+                  @Override public void writeTo(BufferedSink sink) throws IOException {
+                    sink.write(buffer, buffer.size());
+                  }
+                })
+                .build();
+          } catch (IOException e) {
+            e.printStackTrace();
```
Comment (artem-zinnatullin) [inline retrofit/src/main/java/retrofit/OkHttpCall.java]: Looks suspicious, if we can not read request's body due to `IOException` (bad connection, etc), this exception should be thrown up to the caller, shouldn't it?

**Comment by artem-zinnatullin** [comment]
No tests :crying_cat_face:

File: retrofit/src/main/java/retrofit/OkHttpCall.java (hunk 1 above)
Comment (JakeWharton) [inline retrofit/src/main/java/retrofit/OkHttpCall.java]: never for locals

**Comment by JakeWharton** [comment]
This is not the direction in which we want logging to go.

HTTP-level logging belongs in an interceptor. Whether we provide one in a sibling module with semantics similar to V1 remains to be seen.

As to argument and response value `toString` logging, that's probably something we should provide, but I haven't figured out the interface to support it yet. That can be discussed in #378 which is what is tracking logging in general.

**Comment by f2prateek** [comment]
I considered using interceptors, but the problem I ran into was that it would have to be added first to the list of interceptors, which might not always be possible when people set their own client.

Maybe we could ship a logging interceptor that people have to manually set on the client?

**Comment by JakeWharton** [comment]
Why does it have to be first? If anything I would think you'd want it last to see the actual HTTP data that's going across the wire.

**Comment by f2prateek** [comment]
Interesting, I was thinking that Retrofit should only log the request as it transformed it for OkHttp, and the response as it received from OkHttp before it ran it through the converter, instead of worrying about the work that other interceptors might have done.

By running last, other interceptors will have transformed the original request made by Retrofit into something different by the time it was sent across the wire, or rewritten a response into something else by the time Retrofit receives it.

**Comment by JakeWharton** [comment]
One of the complaints that came up frequently from V1 was the fact that the
logging was a fake representation of the HTTP request and response. It was
a best guess, but often did not represent what was actually sent along the
wire. Providing an interceptor, though, would allow the person to choose.

**Pull request closed without merging**